File: typescript/src/context.ts
export type TaskEventType = "record" | "checkpoint" | "log" | "metric"

export type TaskEvent = {
  runId: string
  taskName: string
  taskVersion: string
  attemptNo: number
  seq: number
  eventType: TaskEventType
  payload: Record<string, unknown>
}

export type DataUpsertPayload = {
  runId: string
  taskName: string
  taskVersion: string
  attemptNo: number
  dataset: string
  rows: Record<string, unknown>[]
  keyFields: string[]
}

export type DataQueryPayload = {
  runId: string
  taskName: string
  taskVersion: string
  attemptNo: number
  dataset: string
  fields: string[]
  filters: Record<string, unknown>
  limit: number
}

export type UpsertFn = (payload: DataUpsertPayload) => Promise<void> | void
export type QueryFn = (payload: DataQueryPayload) => Promise<Record<string, unknown>[]> | Record<string, unknown>[]
export type CapabilityProvider = (name: string) => unknown

export interface TaskContext {
  emit(schema: string, record: Record<string, unknown>): void
  checkpoint(data: Record<string, unknown>): void
  log(level: string, msg: string, fields?: Record<string, unknown>): void
  metric(name: string, value: number, tags?: Record<string, string>): void
  capability(name: string): unknown
  cap(): (name: string) => unknown
  cap(name: string): unknown
  upsert(dataset: string, rows: Record<string, unknown>[], keyFields: string[]): Promise<void>
  query(
    dataset: string,
    fields: string[],
    filters: Record<string, unknown>,
    limit: number,
  ): Promise<Record<string, unknown>[]>
}

type CollectorContextOptions = {
  runId: string
  taskName: string
  taskVersion: string
  attemptNo?: number
  eventSink?: (event: TaskEvent) => void
  upsertFn?: UpsertFn
  queryFn?: QueryFn
  capabilityProvider?: CapabilityProvider
}

const noopEventSink = (_event: TaskEvent): void => {
  return
}

const noopUpsert: UpsertFn = async (_payload) => {
  return
}

const noopQuery: QueryFn = async (_payload) => {
  return []
}

const noopCapability: CapabilityProvider = (_name) => undefined

export class CollectorContext implements TaskContext {
  private readonly runId: string
  private readonly taskName: string
  private readonly taskVersion: string
  private readonly attemptNo: number
  private readonly eventSink: (event: TaskEvent) => void
  private readonly upsertSink: UpsertFn
  private readonly querySink: QueryFn
  private readonly capabilitySink: CapabilityProvider
  private seq = 0

  constructor(options: CollectorContextOptions) {
    this.runId = options.runId
    this.taskName = options.taskName
    this.taskVersion = options.taskVersion
    this.attemptNo = options.attemptNo ?? 1
    this.eventSink = options.eventSink ?? noopEventSink
    this.upsertSink = options.upsertFn ?? noopUpsert
    this.querySink = options.queryFn ?? noopQuery
    this.capabilitySink = options.capabilityProvider ?? noopCapability
  }

  emit(schema: string, record: Record<string, unknown>): void {
    this.pushEvent("record", { schema, record })
  }

  checkpoint(data: Record<string, unknown>): void {
    this.pushEvent("checkpoint", { data })
  }

  log(level: string, msg: string, fields: Record<string, unknown> = {}): void {
    this.pushEvent("log", { level, msg, fields })
  }

  metric(name: string, value: number, tags: Record<string, string> = {}): void {
    this.pushEvent("metric", { name, value, tags })
  }

  capability(name: string): unknown {
    return this.capabilitySink(name)
  }

  cap(): (name: string) => unknown
  cap(name: string): unknown
  cap(name?: string): unknown {
    if (typeof name === "string") {
      return this.capability(name)
    }
    return this.capability.bind(this)
  }

  async upsert(dataset: string, rows: Record<string, unknown>[], keyFields: string[]): Promise<void> {
    await this.upsertSink({
      runId: this.runId,
      taskName: this.taskName,
      taskVersion: this.taskVersion,
      attemptNo: this.attemptNo,
      dataset,
      rows,
      keyFields,
    })
  }

  async query(
    dataset: string,
    fields: string[],
    filters: Record<string, unknown>,
    limit: number,
  ): Promise<Record<string, unknown>[]> {
    return await this.querySink({
      runId: this.runId,
      taskName: this.taskName,
      taskVersion: this.taskVersion,
      attemptNo: this.attemptNo,
      dataset,
      fields,
      filters,
      limit,
    })
  }

  private pushEvent(eventType: TaskEventType, payload: Record<string, unknown>): void {
    this.seq += 1
    this.eventSink({
      runId: this.runId,
      taskName: this.taskName,
      taskVersion: this.taskVersion,
      attemptNo: this.attemptNo,
      seq: this.seq,
      eventType,
      payload,
    })
  }
}

File: python/octoarms_sdk/context.py
from __future__ import annotations

from collections.abc import Callable
from typing import Any


EventSink = Callable[[dict[str, Any]], None]
UpsertFn = Callable[[dict[str, Any]], None]
QueryFn = Callable[[dict[str, Any]], list[dict[str, Any]]]
CapabilityProvider = Callable[[str], Any]


def _noop_event_sink(_: dict[str, Any]) -> None:
    return None


def _noop_upsert(_: dict[str, Any]) -> None:
    return None


def _noop_query(_: dict[str, Any]) -> list[dict[str, Any]]:
    return []


def _noop_capability(_: str) -> Any:
    return None


class TaskContext:
    def __init__(
        self,
        *,
        run_id: str,
        task_name: str,
        task_version: str,
        attempt_no: int = 1,
        event_sink: EventSink | None = None,
        upsert: UpsertFn | None = None,
        query: QueryFn | None = None,
        data_upsert: UpsertFn | None = None,
        data_query: QueryFn | None = None,
        capability_provider: CapabilityProvider | None = None,
    ) -> None:
        self.run_id = run_id
        self.task_name = task_name
        self.task_version = task_version
        self.attempt_no = attempt_no
        self._event_sink = event_sink or _noop_event_sink
        self._data_upsert = upsert or data_upsert or _noop_upsert
        self._data_query = query or data_query or _noop_query
        self._capability_provider = capability_provider or _noop_capability
        self._seq = 0

    def emit(self, schema: str, record: dict[str, Any]) -> None:
        self._push_event("record", {"schema": schema, "record": record})

    def checkpoint(self, data: dict[str, Any]) -> None:
        self._push_event("checkpoint", {"data": data})

    def log(self, level: str, msg: str, fields: dict[str, Any] | None = None) -> None:
        self._push_event("log", {"level": level, "msg": msg, "fields": fields or {}})

    def metric(self, name: str, value: float, tags: dict[str, str] | None = None) -> None:
        self._push_event("metric", {"name": name, "value": value, "tags": tags or {}})

    def capability(self, name: str) -> Any:
        return self._capability_provider(name)

    def cap(self, name: str | None = None) -> Any:
        if name is None:
            return self.capability
        return self.capability(name)

    def upsert(self, dataset: str, rows: list[dict[str, Any]], key_fields: list[str]) -> None:
        self._data_upsert(
            {
                "run_id": self.run_id,
                "task_name": self.task_name,
                "task_version": self.task_version,
                "attempt_no": self.attempt_no,
                "dataset": dataset,
                "rows": rows,
                "key_fields": key_fields,
            }
        )

    def query(
        self,
        dataset: str,
        fields: list[str],
        filters: dict[str, Any],
        limit: int,
    ) -> list[dict[str, Any]]:
        return self._data_query(
            {
                "run_id": self.run_id,
                "task_name": self.task_name,
                "task_version": self.task_version,
                "attempt_no": self.attempt_no,
                "dataset": dataset,
                "fields": fields,
                "filters": filters,
                "limit": limit,
            }
        )

    def _push_event(self, event_type: str, payload: dict[str, Any]) -> None:
        self._seq += 1
        self._event_sink(
            {
                "run_id": self.run_id,
                "task_name": self.task_name,
                "task_version": self.task_version,
                "attempt_no": self.attempt_no,
                "seq": self._seq,
                "event_type": event_type,
                "payload": payload,
            }
        )
